### PR TITLE
Restrict notifications

### LINKED ISSUE #1707: System.Runtime.GetNotifications modifies original stack item
If `Notify` is invoked with array argument, it is passed without copy. Thus is possible to modify original stack item by modifying result of `GetNotifications`. I think it is a bug, as I expect `Notify` arguments to be "freezed" at the moment I call `Notify`.

Following test fails:
```csharp
[TestMethod]
public void Runtime_GetNotifications_Test()
{
        var snapshot = Blockchain.Singleton.GetSnapshot();
        using (var engine = new ApplicationEngine(TriggerType.Application, null, snapshot, 0, true))
        using (var script = new ScriptBuilder())
        {
            // Notification
            script.EmitPush("eventPayload");
            script.EmitPush(1);
            script.Emit(OpCode.PACK);
            script.Emit(OpCode.DUP); // save array
            script.EmitPush("eventName");
            script.EmitSysCall(ApplicationEngine.System_Runtime_Notify);
            script.Emit(OpCode.PUSHNULL);
            script.EmitSysCall(ApplicationEngine.System_Runtime_GetNotifications);
            engine.LoadScript(script.ToArray());
            Assert.AreEqual(VMState.HALT, engine.Execute());
            Assert.AreEqual(2, engine.ResultStack.Count);
            Assert.AreEqual(1, engine.Notifications.Count);
            var array = (VM.Types.Array)engine.ResultStack.Pop();
            var notif = (VM.Types.Array)array[0];
            var item = (VM.Types.Array)notif[2]; // array with a single "testEvent" string
            Assert.AreEqual("eventPayload", item[0].GetString());
            
            item[0] = "modifiedEvent"; // modity result of GetNotifications
            array = (VM.Types.Array)engine.ResultStack.Pop(); // load previously savedarray
            Assert.AreEqual("eventPayload", array[0].GetString());
        }
}
```

### PR DESCRIPTION
**Pull request by erikzhang**
Close #1707

### CODE REVIEW

File: src/neo/SmartContract/ApplicationEngine.Runtime.cs [hunk 1]
```diff
@@ -151,10 +151,11 @@ internal void RuntimeNotify(byte[] eventName, Array state)
         {
             if (eventName.Length > MaxEventName) throw new ArgumentException();
             if (!CheckItemForNotification(state)) throw new ArgumentException();
-            SendNotification(CurrentScriptHash, Encoding.UTF8.GetString(eventName), state);
+            if (state.Any(p => !p.IsNull && !(p is PrimitiveType))) throw new ArgumentException();
```
Comment (shargon) [inline src/neo/SmartContract/ApplicationEngine.Runtime.cs:154]: we can't send an array?

Comment (roman-khimov) [reply src/neo/SmartContract/ApplicationEngine.Runtime.cs:154]: I don't think it's worth restricting this interface that much, we should just make a deep copy of notification's stack item and make a copy again if anyone is to retreive it via `GetNotifications`, that's the behavior I would expect.

Comment (shargon) [reply src/neo/SmartContract/ApplicationEngine.Runtime.cs:154]: > I don't think it's worth restricting this interface that much, we should just make a deep copy of notification's stack item and make a copy again if anyone is to retreive it via GetNotifications, that's the behavior I would expect.

Agree

Comment (erikzhang) [reply src/neo/SmartContract/ApplicationEngine.Runtime.cs:154]: Okay, I will modify it.

Comment (erikzhang) [reply src/neo/SmartContract/ApplicationEngine.Runtime.cs:154]: > I don't think it's worth restricting this interface that much, we should just make a deep copy of notification's stack item and make a copy again if anyone is to retreive it via GetNotifications, that's the behavior I would expect.

The first copy is not necessary, right?

Comment (roman-khimov) [reply src/neo/SmartContract/ApplicationEngine.Runtime.cs:154]: > The first copy is not necessary, right?

I think it is necessary, when a contract emits something it must be immutable, like a log line, when you're logging something you're not supposed to change this line afterwards, so there shouldn't be any possibility of doing that. And notifications are a part of contract's API, so any receivers of these notifications (smart contract backends mostly) should have some guarantees of them being intact, delivered exactly as the contract emitted them.

Comment (erikzhang) [reply src/neo/SmartContract/ApplicationEngine.Runtime.cs:154]: But the only chance you can modify it is `GetNotifications`.

**Comment by erikzhang** [comment]
Wait for https://github.com/neo-project/neo-vm/pull/335

File: src/neo/SmartContract/ApplicationEngine.Runtime.cs (hunk 1 above)
Comment (roman-khimov) [reply src/neo/SmartContract/ApplicationEngine.Runtime.cs:154]: I don't think so, there also a possibility for things like this:
```
[TestMethod]
public void Runtime_GetNotifications_Test()
{
        var snapshot = Blockchain.Singleton.GetSnapshot();
        using (var engine = new ApplicationEngine(TriggerType.Application, null, snapshot, 0, true))
        using (var script = new ScriptBuilder())
        {
            // Notification                                                                                                                                                                                                                                                                                                                                                                  
            script.EmitPush("eventPayload");
            script.EmitPush(1);
            script.Emit(OpCode.PACK);
            script.Emit(OpCode.DUP); // save array                                                                                                                                                                                                                                                                                                                                           
            script.EmitPush("eventName");
            script.EmitSysCall(ApplicationEngine.System_Runtime_Notify);
            script.EmitPush(0);
            script.EmitPush("modifiedEvent");
            script.Emit(OpCode.SETITEM);
            engine.LoadScript(script.ToArray());
            Assert.AreEqual(VMState.HALT, engine.Execute());
            Assert.AreEqual(1, engine.Notifications.Count);
            var array = (VM.Types.Array)engine.Notifications[0].State;
            Assert.AreEqual("eventPayload", array[0].GetString());
        }
}
```
And at the moment it works this way:
```
   Assert.AreEqual failed. Expected:<eventPayload>. Actual:<modifiedEvent>. 
```
Which I think is not correct and even dangerous (this array may be passed between contracts, for example).

**Comment by shargon** [comment]
Close?

**Comment by erikzhang** [comment]
Ready to merge.

File: src/neo/SmartContract/NotifyEventArgs.cs [hunk 2]
```diff
@@ -29,7 +29,12 @@ public void FromStackItem(StackItem stackItem)
 
         public StackItem ToStackItem(ReferenceCounter referenceCounter)
         {
-            return new Array(referenceCounter) { ScriptHash.ToArray(), EventName, State };
+            return new Array(referenceCounter)
+            {
+                ScriptHash.ToArray(),
+                EventName,
+                State.DeepCopy()
```
Comment (shargon) [inline src/neo/SmartContract/NotifyEventArgs.cs:36]: Maybe only with this it's enough? We make two deep copy per notifications.

Comment (erikzhang) [reply src/neo/SmartContract/NotifyEventArgs.cs:36]: We have to make two deep copy. Otherwise the sender can modify it after sent.

**Pull request merged**